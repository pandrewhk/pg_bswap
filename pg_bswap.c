/*-------------------------------------------------------------------------
 *
 * pg_bswap.c
 *		byte order reversal function for PostgreSQL
 *
 * Copyright (c) 2022-2023, Andrew Pantyukhin <infofarmer@FreeBSD.org>
 *
 * IDENTIFICATION
 *		pg_bswap/pg_bswap.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"
#include "fmgr.h"
#include "utils/lsyscache.h"
#include "port/pg_bswap.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(bswap);
Datum
bswap(PG_FUNCTION_ARGS) {
	Oid		typ = get_fn_expr_argtype(fcinfo->flinfo, 0);
	int16	typlen;
	bool	typbyval;

	if (!OidIsValid(typ))
		elog(ERROR, "could not determine data type of input");

	get_typlenbyval(typ, &typlen, &typbyval);

	//pass by reference, variable-length
	if (typlen == -1){
		struct varlena *d = PG_DETOAST_DATUM_COPY(PG_GETARG_VARLENA_PP(0));
		unsigned char *ptr = (unsigned char *) VARDATA_ANY(d);
		int32 dataLen = VARSIZE_ANY_EXHDR(d);

		unsigned char *start, *end;

		for (start = ptr, end = ptr + dataLen - 1; start < end; ++start, --end) {
			unsigned char swap = *start;
			*start = *end;
			*end = swap;
		}
		PG_RETURN_POINTER(d);
	}

	//pass by reference, fixed length
	if (typlen > 0 && !typbyval) {
		unsigned char *d = (unsigned char *) PG_GETARG_POINTER(0);
		unsigned char *r = (unsigned char *) palloc(typlen);

		for (int16 i = 0; i < typlen; ++i)
			*(r+i) = *(d+typlen-i-1); //copy backwards

		PG_RETURN_POINTER(r);
	}

	//pass by value, fixed length
	if (typbyval && typlen == 1)
		PG_RETURN_DATUM(PG_GETARG_DATUM(0));

	if (typbyval && typlen == 2)
		PG_RETURN_UINT16(pg_bswap16(PG_GETARG_UINT16(0)));

	if (typbyval && typlen == 4)
		PG_RETURN_UINT32(pg_bswap32(PG_GETARG_UINT32(0)));

	if (typbyval && typlen == 8)
		PG_RETURN_INT64(pg_bswap64(PG_GETARG_INT64(0)));

	elog(ERROR, "unexpected argument type");
	PG_RETURN_NULL();
}
