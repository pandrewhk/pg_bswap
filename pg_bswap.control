comment = 'byte swap for various data types in PostgreSQL'
default_version = '0.1'
module_pathname = '$libdir/pg_bswap'
relocatable = true
