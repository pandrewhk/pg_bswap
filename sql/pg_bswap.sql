CREATE EXTENSION pg_bswap;
-- val fixed
SELECT bswap(1::bool);
SELECT bswap(1::char);
SELECT bswap(1::int2);
SELECT bswap(1::int4);
SELECT bswap(1::int8);
SELECT bswap(-2::int2);
SELECT bswap(-2::int4);
SELECT bswap(-2::int8);
-- ref varlena
SELECT bswap('abc'::text);
SELECT bswap('123'::json);
-- ref fixed
select bswap(('a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11')::uuid);
