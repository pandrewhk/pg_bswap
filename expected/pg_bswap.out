CREATE EXTENSION pg_bswap;
-- val fixed
SELECT bswap(1::bool);
 bswap 
-------
 t
(1 row)

SELECT bswap(1::char);
 bswap 
-------
 1
(1 row)

SELECT bswap(1::int2);
 bswap 
-------
   256
(1 row)

SELECT bswap(1::int4);
  bswap   
----------
 16777216
(1 row)

SELECT bswap(1::int8);
       bswap       
-------------------
 72057594037927936
(1 row)

SELECT bswap(-2::int2);
 bswap 
-------
  -257
(1 row)

SELECT bswap(-2::int4);
   bswap   
-----------
 -16777217
(1 row)

SELECT bswap(-2::int8);
       bswap        
--------------------
 -72057594037927937
(1 row)

-- ref varlena
SELECT bswap('abc'::text);
 bswap 
-------
 cba
(1 row)

SELECT bswap('123'::json);
 bswap 
-------
 321
(1 row)

-- ref fixed
select bswap(('a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11')::uuid);
                bswap                 
--------------------------------------
 110a38bd-b96b-6dbb-f84e-0b9c99bceea0
(1 row)

